### PR TITLE
feat(snapshots): Add min-retention-days feature

### PR DESCRIPTION
**Pull request by akash-rp**
This feature adds an option "min-retention-days" to policy. This option when set greater than 0 keeps all the snapshots within provided days. 

For example:
If we set "min-retention-days" to 14. All the snapshots within 14 days will be kept. The snapshots behind 14 days will be expired

### CODE REVIEW

File: cli/command_policy_set_retention.go
Comment (jkowalski) [inline cli/command_policy_set_retention.go:28]: can we call this `min-retention-days` everywhere and rename . The word "within" is not obvious.

Comment (akash-rp) [reply cli/command_policy_set_retention.go:28]: Okay

File: snapshot/policy/retention_policy.go
Comment (jkowalski) [inline snapshot/policy/retention_policy.go:151]: i don't think this logic belongs here, you should have a separate for loop, this will avoid a bunch of hacks in the current loop.

Comment (akash-rp) [reply snapshot/policy/retention_policy.go:151]: What changes do I need to make

**Comment by akash-rp** [comment]
Please approve pending workflows

File: cli/command_policy_set_retention.go
Comment (adowair) [inline cli/command_policy_set_retention.go:28]:
```suggestion
	cmd.Flag("min-retention-days", "Minimum number of days to keep per source (or 'inherit)").PlaceHolder("N").StringVar(&c.policySetKeepMinDays)
```

File: snapshot/policy/retention_policy.go
Comment (adowair) [inline snapshot/policy/retention_policy.go:173]: Should this say "min"?
```suggestion
				keepReasons = []string{"Inside min retention days"}
```

**Review by adowair — COMMENTED** [review]
Hey @akash-rp, welcome to the Kopia community! I'm still not sure exactly what the feature entails (the verbiage in the PR isn't consistent). Will you please update the PR definition with a summary of the change/usage/effects? I have a couple of nits/questions below.  Thanks for putting this PR up!

File: snapshot/policy/retention_policy.go
Comment (akash-rp) [reply snapshot/policy/retention_policy.go:173]: Yes, that should be min

File: snapshot/policy/retention_policy.go
Comment (adowair) [inline snapshot/policy/retention_policy.go:173]: Should we append the reason here instead? There may be an overlap in the reasons for which a snapshot qualifies to be retained. Since `within` is the last member of `cases` struct, this won't get clobbered right now, but what if the order of members in `cases` struct changes in the future?

File: snapshot/policy/retention_policy.go
Comment (adowair) [inline snapshot/policy/retention_policy.go:171]: Nit: is it necessary to compare to zero here? What happens if we remove lines `154-155` and do this:
```suggestion
		if r.KeepMinDays != nil {
```

**Review by adowair — CHANGES_REQUESTED** [review]
Thanks @akash-rp, I've made some suggestions below

File: snapshot/policy/retention_policy.go
Comment (adowair) [reply snapshot/policy/retention_policy.go]: I've left some suggestions below. Alternatively, (and probably more correctly), this logic should be pulled out of this loop/function completely. See #660 for a similar change

File: snapshot/policy/retention_policy.go
Comment (akash-rp) [reply snapshot/policy/retention_policy.go:177]: I used `c.max = r.KeepMinDays` as the variable `max` already exists, Or else I think we need to add a new variable to the `case `struct. Can you suggest any changes

Comment (adowair) [reply snapshot/policy/retention_policy.go]: I believe that cutoff time here will serve your purpose (do double check that though), it will retain all snapshots within the calculated period, and if `c.max` is `MAX_INT`, it'll do so for any number of such snapshots.

File: snapshot/policy/retention_policy.go
Comment (akash-rp) [reply snapshot/policy/retention_policy.go]: I have removed this logic from loop.

File: snapshot/policy/retention_policy.go
Comment (akash-rp) [reply snapshot/policy/retention_policy.go]: I have moved this logic out of `for` loop

**Comment by akash-rp** [comment]
@jkowalski Hello, can you review changes

File: cli/command_policy_set_retention.go
Comment (jkowalski) [inline cli/command_policy_set_retention.go:43]: please fix the typo "minimun", also rename the fields to match the flags.

the description of the flag is too similar to "number of daily backups to keep". 
How about "minimum number of days to retain the snapshot after creation" or something to that effect

File: snapshot/policy/retention_policy.go
Comment (jkowalski) [inline snapshot/policy/retention_policy.go:160]: you wouldn't need this if you simply remove entry from `cases` and handle it separately.

File: snapshot/policy/policy_tree.go
Comment (jkowalski) [inline snapshot/policy/policy_tree.go:49]: nil is a valid option here, you can leave this uninitialized.

File: snapshot/policy/retention_policy.go
Comment (akash-rp) [reply snapshot/policy/retention_policy.go:155]: Here we need to pass the variable `maxCompleteStartTime` to this function which is not available now

**Comment by akash-rp** [comment]
@jkowalski please look into this

**Comment by jkowalski** [comment]
this definitely needs tests before we can merge it

**Comment by akash-rp** [comment]
Okay. Is their any ETA

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by adowair** [comment]
@akash-rp for any new feature, I think we'd normally expect tests for the feature to be included in the same pull request. If you take a look at `snapshots/policy/retention_policy_test.go`, there are some tests there that look like they could be a good starting point for you. Let us know if you have questions 😄

**Comment by adowair** [comment]
Thanks for the tests @akash-rp, it looks like this code won't compile though:
```shell
ali.dowair -> ~/dev/kopia @ (retention-improvements) $ make ci-tests
...
# github.com/kopia/kopia/cli [github.com/kopia/kopia/cli.test]
cli/command_policy_show.go:146:164: syntax error: unexpected newline, expecting comma or )
cli/command_policy_show.go:147:3: syntax error: unexpected newline, expecting comma or }
cli/command_policy_show.go:148:2: syntax error: unexpected newline, expecting comma or )
cli/command_policy_show.go:150:27: syntax error: unexpected string at end of statement
# github.com/kopia/kopia/cli
cli/command_policy_show.go:146:164: syntax error: unexpected newline, expecting comma or )
cli/command_policy_show.go:147:3: syntax error: unexpected newline, expecting comma or }
cli/command_policy_show.go:148:2: syntax error: unexpected newline, expecting comma or )
cli/command_policy_show.go:150:27: syntax error: unexpected string at end of statement
make: *** [check-locks] Error 2
```

Please ensure that the code compiles (`make kopia`) and that the test runs successfully (`make ci-tests`). There's a short [contribution guideline page](https://kopia.io/docs/contribution-guidelines/) that'll be helpful.

**Comment by akash-rp** [comment]
Fixed Error.

**Comment by akash-rp** [comment]
@adowair I have cleared errors.

File: cli/command_policy_show.go
Comment (jkowalski) [inline cli/command_policy_show.go:146]: You have a typo in the name: perhaps use `Min. Retention Days`

File: snapshot/policy/retention_policy.go
Comment (jkowalski) [inline snapshot/policy/retention_policy.go:31]: json fields should be lowercase

File: snapshot/policy/retention_policy.go
Comment (jkowalski) [inline snapshot/policy/retention_policy.go:43]: same here

File: snapshot/policy/retention_policy.go
Comment (jkowalski) [inline snapshot/policy/retention_policy.go:121]: this is not correct, the value returns the "count of snapshots" not the "number of snapshot + number of days" which does not really make sense. I think reverting this line is fine.

File: snapshot/policy/retention_policy.go
Comment (jkowalski) [inline snapshot/policy/retention_policy.go:156]: something like:

```
if r.MinRetentionDays.OrDefault(0) > 0 && !s.StartTime.Before(cutoff.minDays) {
  keepReasons = append(keepReasons, "min-retention-days")
}
```

File: snapshot/policy/retention_policy.go
Comment (jkowalski) [inline snapshot/policy/retention_policy.go:157]: let's not capitalize the string and should not include spaces.

File: README.md
Comment (jkowalski) [reply README.md:1]: please revert the entire file, this only has invisible changes

File: snapshot/policy/retention_policy.go
Comment (akash-rp) [reply snapshot/policy/retention_policy.go:121]: The problem is, if we don't add `r.MinRetentionDays.OrDefault(0)` and only `Min. Retention days` is set to some value and  other polices is set to zero then effectiveKeepLatest() function returns `keep-latest = maxInt`, so then every snapshot will have retention reasons as `[min-retention-days, latest-n ]` where `n = number of snapshots`

Comment (akash-rp) [reply snapshot/policy/retention_policy.go:121]: @jkowalski Can you suggest something

**Pull request closed without merging**
